### PR TITLE
User Parameter Values Are Unique Per Dashboard

### LINKED ISSUE #43001: "Duplicate Dashboard" messes up last used filter values - one dashboard overrides another
**Context**

We're facing wrong logic for duplicated dashboards. it turned out, when we make a deep clone of the dashboard (duplicate), all filters are cloned with the same ids so currently if I change a main dashboard's filter, duplicated dashboard will be affected as well

logic related to dashboard copy ignores `parameters` and `dashcards` sent from FE, so we can't override it on our side
https://github.com/metabase/metabase/blob/master/src/metabase/api/dashboard.clj#L364-L412

Current logic of setting `last_used_param_values` presumes we have unique filter ids, so some of those places should be fixed
https://github.com/metabase/metabase/pull/40415/files#diff-4191aab4baca759a3e1663cbe6c28268c326d418f224f12717f00be32dd47583R38-R50

https://github.com/metabase/metabase/assets/125459446/e29ffdba-c080-4c1e-b1d8-55778d30c202

caught by CI https://github.com/metabase/metabase/actions/runs/9163281195/job/25192372653?pr=42777

### LINKED ISSUE #44858: Updating filter on one dashboard automatically updates filters on other dashboards with cards referencing the same data
### Describe the bug

Updating filter on one dashboard automatically updates filters on other dashboards with cards referencing the same data. It used to not be the case. 

eg. we have sales rep dashboard for each sales rep with a text filter defaults to that sales rep's name. Recently (not sure since when), whoever opens their dashboard first would change that text filter's value to their name across all sales reps' dashboards. 

### To Reproduce

1. open a dashboard with a text filter applying to a text field of a card. The text filter value defaults to "a" in the dashboard setting.
2. open another dashboard with a text filter applying to the same text field of a different card. The text filter value defaults to "b" in the dashboard setting.
3. open the dashboard from step 1, and the text filter value now defaults to "b" instead of "a".

### Expected behavior

changing filters in one dashboard should not change filters automatically of other dashboards

### Logs

_No response_

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.23+9",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.23",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.23+9",
    "os.name": "Linux",
    "os.version": "5.10.217-205.860.amzn2.x86_64",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "postgres",
      "snowflake"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "13.13"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.7.3"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2024-06-19",
      "tag": "v0.50.6",
      "hash": "a5fbebf"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

### Severity

blocking some users

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by adam-james-v**
Fixes: #43001
Fixes: #44858

User Parameter Values previously stored just the `parameter-id` and `user-id` as data to look up the last used parameter value. This isn't sufficient as the parameter id is not guaranteed to be unique, in particular this is true when a dashboard is duplicated: the parameters are copied to the new dashboard without changing the parameter id at all.

This means that we need to also store the dashboard-id in the user_parameter_value table and use that to update/get the last used value.

The migration removes existing entries to the user_parameter_value table because I want a non-nullable constraint on the dashboard_id column, and existing entries will not have a value. The only way to backfill those values would be to look through every dashboard containing parameters, and for every parameter check for a matching ID. 

Even if you can do this, there's no way to disambiguate if the same paramter exists on 2 dashboards anyway, so one of them might be wrong. I think it's not worth the trouble considering that removing the entries in this table doesn't break anything; it just becomes a mild inconvenience that a user has to select a filter value again (since the dashboard will use the default value). This was the behaviour of filters in the app until 50 anyway, so for those couple of times where the app's behaviour is regressed, it shouldn't be too shocking.

### CODE REVIEW

File: src/metabase/models/user_parameter_value.clj [hunk 1]
```diff
@@ -52,17 +56,19 @@
...
+          last-used-values  (fn [dashboard-id]
+                              (when (seq all-parameter-ids)
+                                (into {}
+                                      (t2/select-fn-vec
+                                       (fn [{:keys [parameter_id value]}]
+                                         [parameter_id value])
+                                       :model/UserParameterValue
+                                       :user_id user-id
+                                       :dashboard_id dashboard-id
+                                       :parameter_id [:in all-parameter-ids]))))]
```
Comment (escherize) [inline src/metabase/models/user_parameter_value.clj:69]: It is more performant to use this `[model & column]` form, and it should return the same thing:

```suggestion
                                      (t2/select
                                       [:model/UserParameterValue :parameter_id :value]
                                       :user_id user-id
                                       :dashboard_id dashboard-id
                                       :parameter_id [:in all-parameter-ids]))))]
```

File: src/metabase/models/user_parameter_value.clj [hunk 2]
```diff
@@ -33,17 +33,21 @@
...
    value]
   (if value
     (or (pos? (t2/update! :model/UserParameterValue {:user_id      user-id
+                                                     :dashboard_id dashboard-id
                                                      :parameter_id parameter-id}
                           {:value value}))
         (t2/insert! :model/UserParameterValue {:user_id      user-id
+                                               :dashboard_id dashboard-id
                                                :parameter_id parameter-id
                                                :value        value}))
```
[the remark this answers is not in the document]
Comment (escherize) [reply src/metabase/models/user_parameter_value.clj:47]: more info: https://github.com/metabase/toucan/issues/17

**Review by escherize — APPROVED** [review]
Looks good!

File: src/metabase/models/user_parameter_value.clj (hunk 1 above)
Comment (adam-james-v) [reply src/metabase/models/user_parameter_value.clj]: actually, the shape's not quite right so I think I'll keep it the way it was. 
Might look at alternatives to see if there's something more performant that still works the way I need.

**Pull request merged**